Integrations/src/main/java/io/deephaven/integrations/python/PythonReplayListenerAdapter.java

```diff
@@ -69,18 +77,36 @@ public void replay() {
         final TableUpdate update =
                 new TableUpdateImpl(source.getRowSet(), emptyRowSet, emptyRowSet, emptyShift, emptyColumnSet);
         final boolean isReplay = true;
-        pyCallable.call("__call__", update, isReplay);
+        pyListenerCallable.call("__call__", update, isReplay);
     }
 
     @Override
     public void onUpdate(final TableUpdate update) {
         final boolean isReplay = false;
-        pyCallable.call("__call__", update, isReplay);
+        pyListenerCallable.call("__call__", update, isReplay);
+    }
+
+    @Override
+    public void onFailureInternal(Throwable originalException, Entry sourceEntry) {
+        if (!pyOnFailureCallback.isNone()) {
+            try {
+                pyOnFailureCallback.call("__call__", ExceptionUtils.getStackTrace(originalException));
+            } catch (Throwable e) {
+                // If the Python onFailure callback fails, log the new exception
+                // and continue with the original exception.
+                log.error().append("Python on_error callback failed: ").append(e).endl();
+            }
+        }
+        super.onFailureInternal(originalException, sourceEntry);
     }
 
     @Override
     public boolean canExecute(final long step) {
         return super.canExecute(step)
                 && (dependencies.length == 0 || Arrays.stream(dependencies).allMatch(t -> t.satisfied(step)));
     }
+
+    public boolean isFailed() {
+        return failed;
+    }
```

Comment:
Let's put this in the parent class (where `isFailed`) lives. That way we can keep `isFailed` private, since we don't expect to mutate it in the child classes.